File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.kt
class Solution {
    fun longestPalindrome(s: String): String {
        var left = 0
        var maxLen = 0
        
        fun extend(low: Int, high: Int){
            var i = low
            var j = high

            while (i >=0 && j < s.length && s[i] == s[j]){
                i--
                j++
            }

            if (maxLen < j - i -1){
                maxLen = j - i -1
                left = i + 1
            }
        }

        if (s.length < 2){
            return s
        }

        for (i in 0 until s.length -1){
            extend(i, i + 1)
            extend(i, i + 2)
        }

        return s.substring(left, left + maxLen)
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        left = 0
        maxLen = 0

        def check(start, end):
            nonlocal maxLen, left 

            while start >= 0 and end < len(s) and s[start] == s[end]:
                start -= 1
                end += 1
            
            if (end - start - 1) > maxLen:
                maxLen = (end - start - 1)
                left = start + 1
        
        if len(s) <= 1:
            return s

        for i in range(len(s)-1):
            check(i, i+1)
            check(i, i+2)
        return s[left:left + maxLen]
            
            